File: src/main/java/mediathek/gui/dialog/MemoryMonitorDialog.java
package mediathek.gui.dialog;

import mediathek.mainwindow.MemoryUsagePanel;
import mediathek.tool.ApplicationConfiguration;
import org.apache.commons.configuration2.Configuration;
import org.apache.commons.configuration2.sync.LockMode;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ComponentAdapter;
import java.awt.event.ComponentEvent;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.time.Duration;
import java.util.Optional;

public final class MemoryMonitorDialog extends JDialog {
    private static final Duration HISTORY_WINDOW = Duration.ofMinutes(2);
    private static final Duration SAMPLE_INTERVAL = Duration.ofSeconds(1);
    private static final Dimension DEFAULT_SIZE = new Dimension(480, 240);

    private final Configuration configuration = ApplicationConfiguration.getConfiguration();
    private final MemoryUsagePanel memoryUsagePanel = new MemoryUsagePanel(HISTORY_WINDOW, SAMPLE_INTERVAL);

    public MemoryMonitorDialog(@NotNull JFrame parent) {
        super(parent, "Speicherverbrauch", false);

        setType(Type.UTILITY);
        setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);

        memoryUsagePanel.setPreferredSize(DEFAULT_SIZE);
        add(memoryUsagePanel, BorderLayout.CENTER);
        pack();
        restoreBounds();

        addComponentListener(new ComponentAdapter() {
            @Override
            public void componentResized(ComponentEvent event) {
                storeBounds();
            }

            @Override
            public void componentMoved(ComponentEvent event) {
                storeBounds();
            }
        });
        addWindowListener(new WindowAdapter() {
            @Override
            public void windowOpened(WindowEvent event) {
                storeVisibility(true);
            }

            @Override
            public void windowClosed(WindowEvent event) {
                storeVisibility(false);
            }
        });
    }

    @Override
    public void dispose() {
        memoryUsagePanel.close();
        super.dispose();
    }

    private void restoreBounds() {
        readStoredBounds().ifPresentOrElse(this::applyBounds, this::applyDefaultBounds);
    }

    private Optional<DialogBounds> readStoredBounds() {
        configuration.lock(LockMode.READ);
        try {
            var width = configuration.getInt(ApplicationConfiguration.MemoryMonitorDialog.WIDTH, -1);
            var height = configuration.getInt(ApplicationConfiguration.MemoryMonitorDialog.HEIGHT, -1);
            var x = configuration.getInt(ApplicationConfiguration.MemoryMonitorDialog.X, Integer.MIN_VALUE);
            var y = configuration.getInt(ApplicationConfiguration.MemoryMonitorDialog.Y, Integer.MIN_VALUE);

            if (width <= 0 || height <= 0 || x == Integer.MIN_VALUE || y == Integer.MIN_VALUE) {
                return Optional.empty();
            }

            return Optional.of(new DialogBounds(x, y, width, height));
        } finally {
            configuration.unlock(LockMode.READ);
        }
    }

    private void applyBounds(DialogBounds bounds) {
        setBounds(bounds.x(), bounds.y(), bounds.width(), bounds.height());
    }

    private void applyDefaultBounds() {
        setSize(DEFAULT_SIZE);
        setLocationRelativeTo(getOwner());
    }

    private void storeBounds() {
        if (!isShowing()) {
            return;
        }

        var bounds = getBounds();
        configuration.lock(LockMode.WRITE);
        try {
            configuration.setProperty(ApplicationConfiguration.MemoryMonitorDialog.X, bounds.x);
            configuration.setProperty(ApplicationConfiguration.MemoryMonitorDialog.Y, bounds.y);
            configuration.setProperty(ApplicationConfiguration.MemoryMonitorDialog.WIDTH, bounds.width);
            configuration.setProperty(ApplicationConfiguration.MemoryMonitorDialog.HEIGHT, bounds.height);
        } finally {
            configuration.unlock(LockMode.WRITE);
        }
    }

    private void storeVisibility(boolean visible) {
        configuration.setProperty(ApplicationConfiguration.MemoryMonitorDialog.VISIBLE, visible);
    }

    private record DialogBounds(int x, int y, int width, int height) {
    }
}

File: src/main/kotlin/mediathek/gui/dialog/MemoryMonitorDialog.kt
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package mediathek.gui.dialog

import mediathek.config.application.ApplicationConfiguration
import mediathek.mainwindow.MemoryUsagePanel
import java.awt.BorderLayout
import java.awt.Dimension
import java.awt.event.ComponentAdapter
import java.awt.event.ComponentEvent
import java.awt.event.WindowAdapter
import java.awt.event.WindowEvent
import java.time.Duration
import javax.swing.JDialog
import javax.swing.JFrame

class MemoryMonitorDialog(
    parent: JFrame,
    private val onClose: Runnable,
) : JDialog(parent, "Speicherverbrauch", false) {

    private val memoryUsagePanel = MemoryUsagePanel(HISTORY_WINDOW, SAMPLE_INTERVAL)
    private var preserveVisibilityOnClose = false

    init {
        type = Type.UTILITY
        defaultCloseOperation = DISPOSE_ON_CLOSE

        memoryUsagePanel.preferredSize = DEFAULT_SIZE
        add(memoryUsagePanel, BorderLayout.CENTER)
        pack()
        restoreBounds()

        addComponentListener(object : ComponentAdapter() {
            override fun componentResized(event: ComponentEvent) {
                storeBounds()
            }

            override fun componentMoved(event: ComponentEvent) {
                storeBounds()
            }
        })
        addWindowListener(object : WindowAdapter() {
            override fun windowOpened(event: WindowEvent) {
                storeVisibility(true)
            }

            override fun windowClosed(event: WindowEvent) {
                storeVisibility(preserveVisibilityOnClose)
                notifyClosed()
            }
        })
    }

    override fun dispose() {
        memoryUsagePanel.close()
        super.dispose()
    }

    fun disposeForShutdown() {
        preserveVisibilityOnClose = true
        dispose()
    }

    private fun restoreBounds() {
        readStoredBounds()?.let(::applyBounds) ?: applyDefaultBounds()
    }

    private fun readStoredBounds(): DialogBounds? {
        val state = ApplicationConfiguration.getInstance().memoryMonitorDialogState
        return if (state.hasStoredBounds()) {
            DialogBounds(state.x, state.y, state.width, state.height)
        } else {
            null
        }
    }

    private fun applyBounds(bounds: DialogBounds) {
        setBounds(bounds.x, bounds.y, bounds.width, bounds.height)
    }

    private fun applyDefaultBounds() {
        size = DEFAULT_SIZE
        setLocationRelativeTo(owner)
    }

    private fun storeBounds() {
        if (!isShowing) {
            return
        }

        val bounds = bounds
        ApplicationConfiguration.getInstance().setMemoryMonitorDialogBounds(bounds.x, bounds.y, bounds.width, bounds.height)
    }

    private fun storeVisibility(visible: Boolean) {
        ApplicationConfiguration.getInstance().memoryMonitorDialogVisible = visible
    }

    private fun notifyClosed() {
        onClose.run()
    }

    private data class DialogBounds(
        val x: Int,
        val y: Int,
        val width: Int,
        val height: Int,
    )

    companion object {
        private val HISTORY_WINDOW: Duration = Duration.ofMinutes(2)
        private val SAMPLE_INTERVAL: Duration = Duration.ofSeconds(1)
        private val DEFAULT_SIZE = Dimension(480, 240)
    }
}
